src/clang-cpp-frontend/clang_cpp_convert_vft.cpp

```diff
@@ -0,0 +1,585 @@
+
+    /*
+     * 4. deal with overriding method
+     */
+    if(md->begin_overridden_methods() != md->end_overridden_methods())
+    {
+      /*
+       * Assume it *always* points to one overriden method in base class
+       * TODO: Use a loop if there are more overriden methods
+       *    md->overriden_methods() should do the job
+       */
+      assert(md->size_overridden_methods() == 1);
+      add_thunk_method(*(md->begin_overridden_methods()), comp, type);
+    }
+  }
+
+  /*
+   * Set up virtual function table(vft) variable symbols
+   * Each vft is modelled as a struct of function pointers.
+   */
+  setup_vtable_struct_variables(cxxrd, type);
+
+  return false;
+}
+
+bool clang_cpp_convertert::annotate_virtual_overriding_methods(
+  const clang::CXXMethodDecl *md,
+  struct_typet::componentt &comp)
+{
+  std::string method_id, method_name;
+  clang_c_convertert::get_decl_name(*md, method_name, method_id);
+
+  comp.type().set("#is_virtual", true);
+  comp.type().set("#virtual_name", method_name);
+  comp.set("is_virtual", true);
+  comp.set("virtual_name", method_name);
+
+  return false;
+}
+
+symbolt *
+clang_cpp_convertert::check_vtable_type_symbol_existence(struct_typet &type)
+{
+  irep_idt vt_name = vtable_type_prefix + tag_prefix + type.tag().as_string();
+  return context.find_symbol(vt_name);
+}
+
+symbolt *clang_cpp_convertert::add_vtable_type_symbol(
+  const clang::CXXMethodDecl *md,
+  const struct_typet::componentt &comp,
+  struct_typet &type)
+{
+  /*
+   *  We model the type of the virtual table as a struct type, something like:
+   *   typedef struct {
+   *      void (*do_something)(Base*);
+   *    } VftTag_Base;
+   * Later, we will instantiate a virtual table as:
+   *  VftTag_Base vtable_TagBase@TagBase = { .do_something = &TagBase::do_someting(); }
+   *
+   *  Vtable type has the id in the form of `virtual_table::tag-BLAH`.
+   */
+
+  irep_idt vt_name = vtable_type_prefix + tag_prefix + type.tag().as_string();
+
+  symbolt vt_type_symb;
+  vt_type_symb.id = vt_name;
+  vt_type_symb.name = vtable_type_prefix + type.tag().as_string();
+  vt_type_symb.mode = mode;
+  vt_type_symb.type = struct_typet();
+  vt_type_symb.is_type = true;
+  vt_type_symb.type.set("name", vt_type_symb.id);
+  vt_type_symb.location = comp.location();
+  vt_type_symb.module =
+    get_modulename_from_path(comp.location().file().as_string());
+
+  if(context.move(vt_type_symb))
+  {
+    log_error(
+      "Failed add vtable type symbol {} to symbol table", vt_type_symb.id);
+    abort();
+  }
+
+  return context.find_symbol(vt_name);
+}
+
+void clang_cpp_convertert::add_vptr(
+  const clang::CXXMethodDecl *md,
+  struct_typet &type)
+{
+  /*
+   * We model the virtual pointer as a `component` to the parent class' type.
+   * This will be the vptr pointing to the vtable that contains the overriden functions.
+   *
+   * Vptr has the name in the form of `tag-BLAH@vtable_pointer`.
```

Comment:
what is `BLAH` here? It might be easier on the eye if it was `TYPE_NAME` or similar.

Reply:
Fixed